onnxruntime/core/optimizer/selectors_actions/helpers.cc

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/selectors_actions/actions.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+namespace {
+// Move or remove an edge.
+//   - moves edges from src+src_slot to dest node+dest_slot if provided.
+//   - remove edges for the src+src_slot if dest+dest_slot not provided.
+void ProcessEdge(Graph& graph, Node& src, const InOutDefSlot& src_slot,
+                 Node* dest, const InOutDefSlot* dest_slot) {
+  if (src_slot.in_out == ArgType::kInput) {
+    // move input edge if present
+    auto iter = std::find_if(src.InputEdgesBegin(), src.InputEdgesEnd(),
+                             [&src_slot](const Node::EdgeEnd& edge) {
+                               return (edge.GetDstArgIndex() == src_slot.idx);
+                             });
+
+    // initializer or graph input doesn't have an edge so either zero or one edges to process
+    if (iter != src.InputEdgesEnd()) {
+      const Node& iter_node = iter->GetNode();
+      // need to save this before calling RemoveEdge as that invalidates the iterator
+      auto iter_src_idx = iter->GetSrcArgIndex();
+      graph.RemoveEdge(iter_node.Index(), src.Index(), iter_src_idx, src_slot.idx);
+      if (dest && dest_slot) {
+        graph.AddEdge(iter_node.Index(), dest->Index(), iter_src_idx, dest_slot->idx);
+      }
+    }
+
+  } else {
+    // otherwise we need to move all output edges (if any)
+    auto edges = graph_utils::GraphEdge::GetNodeOutputEdges(src, src_slot.idx);
+    graph_utils::GraphEdge::RemoveGraphEdges(graph, edges);
+    if (dest && dest_slot) {
+      for (const auto& edge : edges) {
+        graph.AddEdge(dest->Index(), edge.dst_node, dest_slot->idx, edge.dst_arg_index);
+      }
+    }
+  }
+}
+
+// move an input or output and its edge between two nodes
+Status MoveInputOutputImpl(Graph& graph, const ValueMoveInfo& move_info, Node& src, Node& dest) {
+  auto& src_defs = (move_info.src_slot.in_out == ArgType::kInput)
+                       ? src.MutableInputDefs()
+                       : src.MutableOutputDefs();
+
+  auto& dest_defs = (move_info.dest_slot.in_out == ArgType::kInput)
+                        ? dest.MutableInputDefs()
+                        : dest.MutableOutputDefs();
+
+  auto process = [&](int src_idx) {
```

Comment:
>int [](http://example.com/codeflow?start=21&length=3)

How about size_t? Then you don't need to cast below and gsl::narrow at the bottom of the function?

Reply:
It needs to be signed in InOutDefSlot. Never mind.